### PR TITLE
feat(module):dmd user id module[PREBID-1]

### PR DESCRIPTION
**Pull request by boppudikarthikc**
## Type of change

- [ ] Bugfix
- [x] Feature
- [ ] New bidder adapter   
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Does this change affect user-facing APIs or examples documented on http://prebid.org?
- [ ] Other

## Description of change

- test parameters for validating bids
```
{
  bidder: '<bidder name>',
  params: {
    // ...
  }
}
```

Be sure to test the integration with your adserver using the [Hello World](/integrationExamples/gpt/hello_world.html) sample page.

- contact email of the adapter’s maintainer
- [ ] official adapter submission

For any changes that affect user-facing APIs or example code documented on http://prebid.org, please provide:

- A link to a PR on the docs repo at https://github.com/prebid/prebid.github.io/

### CODE REVIEW

File: modules/dmdIdSystem.js [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+  /**
+   * decode the stored id value for passing to bid requests
+   * @function decode
+   * @param {(Object|string)} value
+   * @returns {(Object|undefined)}
+   */
+  decode(value) {
+    return value && typeof value['dmd-dgid'] === 'string'
+      ? { 'dmdid': value['dmd-dgid'] }
```
Comment (matthewfitz) [inline modules/dmdIdSystem.js:27]: @boppudikarthikc can you camel case this **dmdId**

```
? { 'dmdId': value['dmd-dgid'] }
```

File: modules/dmdIdSystem.js [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+  getId(config) {
+    try {
+      const configParams = (config && config.params) || {};
+      if (
+        !configParams ||
+        !configParams.apiKey ||
+        typeof configParams.apiKey !== 'string'
+      ) {
+        utils.logError('dmd submodule requires an apiKey.');
+        return;
```
Comment (matthewfitz) [inline modules/dmdIdSystem.js:48]: @boppudikarthikc Can you snake case the configuration for "api_key"

File: modules/dmdIdSystem.js [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+  /**
+   * decode the stored id value for passing to bid requests
+   * @function decode
+   * @param {(Object|string)} value
+   * @returns {(Object|undefined)}
+   */
+  decode(value) {
+    return value && typeof value['dmd-dgid'] === 'string'
+      ? { 'dmdId': value['dmd-dgid'] }
+      : null;
```
Comment (awerlang) [inline modules/dmdIdSystem.js:28]: Here it must return undefined (see jsdoc above).

**Comment by awerlang** [comment]
Review this doc: https://github.com/prebid/Prebid.js/blob/master/PR_REVIEW.md#reviewing-a-new-or-updated-user-id-sub-module

It mentions a couple other things are expected:
* Test coverage;
* References added to 4 files under ```modules/```

**Comment by matthewfitz** [comment]
@boppudikarthikc can you add dmdIdSystem to this submodules registry https://github.com/dmdconnects/Prebid.js/blob/develop/modules/.submodules.json#L23

File: modules/userId/eids.js [hunk 4]
```diff
@@ -96,6 +96,12 @@ const USER_IDS_CONFIG = {
     atype: 3
   },
 
+  // dmdId
+  'dmdid': {
+    source: 'hcn.health',
+    atype: 1
```
Comment (awerlang) [inline modules/userId/eids.js:102]: ```dmdid``` i.e. dmd-dgid cookie is a person-based identifier, thus atype = 3 source: https://github.com/InteractiveAdvertisingBureau/AdCOM/blob/master/AdCOM%20v1.0%20FINAL.md#list_agenttypes

Also, should ```dmdid``` be camel-case?

File: modules/dmdIdSystem.js [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+   */
+  getId(config) {
+    try {
+      const configParams = (config && config.params) || {};
+      if (
+        !configParams ||
+        !configParams.api_key ||
+        typeof configParams.api_key !== 'string'
+      ) {
+        utils.logError('dmd submodule requires an apiKey.');
```
Comment (awerlang) [inline modules/dmdIdSystem.js:47]: typo: apiKey => api_key in the error message.

File: test/spec/modules/dmdIdSystem_spec.js [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+  afterEach(function () {
+    logErrorStub.restore();
+  });
+
+  it('should log an error if no configParams were passed into getId', function () {
+    dmdIdSubmodule.getId();
+    expect(logErrorStub.calledOnce).to.be.true;
+  });
+
+  it('should log an error if configParams doesnot have apiKey passed to getId', function () {
```
Comment (awerlang) [inline test/spec/modules/dmdIdSystem_spec.js:21]: typo: apiKey => api_key

File: test/spec/modules/dmdIdSystem_spec.js [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+    dmdIdSubmodule.getId({});
+    expect(logErrorStub.calledOnce).to.be.true;
+  });
+
+  it('should log an error if configParams has invalid apiKey passed into getId', function () {
```
Comment (awerlang) [inline test/spec/modules/dmdIdSystem_spec.js:26]: typo: apiKey => api_key

File: test/spec/modules/dmdIdSystem_spec.js [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+    dmdIdSubmodule.getId({api_key: 123});
```
Comment (awerlang) [inline test/spec/modules/dmdIdSystem_spec.js:27]: ```getId()``` expects a ```params``` object containing ```api_key```.

File: test/spec/modules/dmdIdSystem_spec.js [hunk 9]
```diff
@@ -0,0 +1,43 @@
+import * as utils from '../../../src/utils.js';
+
+import {dmdIdSubmodule} from 'modules/dmdIdSystem.js';
+
+describe('Dmd ID System', function() {
```
Comment (awerlang) [inline test/spec/modules/dmdIdSystem_spec.js:5]: could you add a positive test case for getId()? Something like ```should call getId() with a valid api_key```

**Review by awerlang — CHANGES_REQUESTED** [review]

**Review by awerlang — DISMISSED** [review]

**Review by awerlang — APPROVED** [review]

**Review by matthewfitz — APPROVED** [review]
Feedback implemented

**Pull request merged**